plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftPageSourceProvider.java

```diff
@@ -0,0 +1,142 @@
+import io.trino.parquet.reader.RowGroupInfo;
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.JdbcSplit;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ConnectorPageSource;
+import io.trino.spi.connector.ConnectorPageSourceProvider;
+import io.trino.spi.connector.ConnectorRecordSetProvider;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorSplit;
+import io.trino.spi.connector.ConnectorTableHandle;
+import io.trino.spi.connector.ConnectorTransactionHandle;
+import io.trino.spi.connector.DynamicFilter;
+import io.trino.spi.connector.RecordPageSource;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.io.MessageColumnIO;
+import org.apache.parquet.schema.MessageType;
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static io.trino.parquet.ParquetTypeUtils.constructField;
+import static io.trino.parquet.ParquetTypeUtils.getColumnIO;
+import static io.trino.parquet.ParquetTypeUtils.getDescriptors;
+import static io.trino.parquet.ParquetTypeUtils.lookupColumnByName;
+import static io.trino.parquet.metadata.PrunedBlockMetadata.createPrunedColumnsMetadata;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_PARQUET_CURSOR_ERROR;
+import static io.trino.plugin.redshift.RedshiftPageSource.handleException;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftPageSourceProvider
+        implements ConnectorPageSourceProvider
+{
+    private final ConnectorRecordSetProvider recordSetProvider;
+    private final TrinoFileSystemFactory fileSystemFactory;
+
+    public RedshiftPageSourceProvider(ConnectorRecordSetProvider recordSetProvider, TrinoFileSystemFactory fileSystemFactory)
+    {
+        this.recordSetProvider = requireNonNull(recordSetProvider, "recordSetProvider is null");
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+    }
+
+    @Override
+    public ConnectorPageSource createPageSource(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableHandle table,
+            List<ColumnHandle> columns,
+            DynamicFilter dynamicFilter)
+    {
+        if (split instanceof JdbcSplit) {
+            return new RecordPageSource(recordSetProvider.getRecordSet(transaction, session, split, table, columns));
+        }
+
+        String path = ((RedshiftUnloadSplit) split).path();
+        Location location = Location.of(path);
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        TrinoInputFile inputFile = fileSystem.newInputFile(location);
```

Comment:
If we UNLOAD with verbose manifest, we can get the file length of each file from parsing it https://docs.aws.amazon.com/redshift/latest/dg/r_UNLOAD_command_examples.html#unload-examples-manifest-verbose
That can be populated into the split and use in `fileSystem.newInputFile(location, length)` to avoid an extra file status call per file.

Reply:
> fileSystem.newInputFile(location, length) to avoid an extra file status call per file..

Good to know. Added file filesize to split